File: 다이나믹 프로그래밍/11054번 가장 긴 바이토닉 부분 수열.cpp
#include<iostream>

#define MAX 1000

using namespace std;

int n, ans = 0;
int arr[MAX];
int dp[MAX], dp_left[MAX], dp_right[MAX];

void input()
{
	cin >> n;
	for (int i = 0; i < n; ++i)
		cin >> arr[i];
}

void solve()
{
	for (int i = 0; i < n; ++i) {
		dp_left[i] = 1;

		for (int j = 0; j < i; ++j)
			if (arr[i] > arr[j])
				dp_left[i] = max(dp_left[i], dp_left[j] + 1);
	}

	for (int i = n - 1; i > -1; --i) {
		dp_right[i] = 1;

		for (int j = n - 1; j > i; --j)
			if (arr[i] > arr[j])
				dp_right[i] = max(dp_right[i], dp_right[j] + 1);
	}

	for (int i = 0; i < n; ++i) {
		dp[i] = dp_left[i] + dp_right[i];

		ans = max(ans, dp[i]);
	}
		
	cout << ans - 1;
}

void init()
{
	ios_base::sync_with_stdio(false);
	cin.tie(0);
	cout.tie(0);
}

int main()
{
	init();
	input();
	solve();

	return 0;
}

File: 다이나믹 프로그래밍/11054번 가장 긴 바이토닉 부분 수열.py
import sys

input = sys.stdin.readline

n = int(input())
arr = list(map(int, input().split()))
dp_increase = [1] * n   # 가장 긴 증가하는 부분 수열의 길이
dp_decrease = [1] * n   # 가장 긴 감소하는 부분 수열의 길이

for i in range(n):
    for j in range(0, i):
        if arr[i] > arr[j]:
            dp_increase[i] = max(dp_increase[i], dp_increase[j] + 1)

for i in range(n - 1, -1, -1):
    for j in range(n - 1, i, -1):
        if arr[i] > arr[j]:
            dp_decrease[i] = max(dp_decrease[i], dp_decrease[j] + 1)

ans = 0

# (가장 긴 증가하는 부분 수열의 길이 + 가장 긴 감소하는 부분 수열의 길이) - 1 = 가장 긴 바이토닉 부분 수열의 길이
for i in range(n):
    ans = max(ans, dp_increase[i] + dp_decrease[i])
print(ans - 1)